Book/php7/zend_engine/zend_opcache.rst

```diff
@@ -1,2 +1,104 @@
+change between requests both the parsing and compilation steps will always yield the same result. Opcache is an
+extension that caches the opcodes and data structures like classes between requests to improve performance. The AST is
+not used after the compilation step (with the exception of constant expressions) and thus does not need to be cached.
+
+Normally, all the memory allocated during the processing of a request gets freed after the request has been processed.
+So to store the compiled opcodes for future requests opcache puts them in a shared memory segment (SHM) that can not
+only be accessed by the same process after the request has finished but also by other processes handling other requests.
+This means that if you are using php-fpm with many children they will only store the opcodes for each PHP file once in
+SHM. 
+
+This comes with a significant restriction: The opcodes and other data structures in SHM must not be mutated by any
+child request (at least not unless these changes should also affect the other processes). If any changes to SHM are
+made there is a locking mechanism to avoid data races.
+
+The classes and OParrays (arrays of opcodes, i.e. functions) in SHM are mostly immutable. There are currently the
+following exceptions:
+
+* ``zend_persistent_script.dynamic_members`` stores information about the state of the cached script
+* ``zend_class_entry.inheritance_cache`` is extended if a new combination of parent class and interfaces is encountered
+* ``zend_op.handler`` can be replaced at runtime by the tracing JIT to start executing JITted code
+* ``zend_string.gc.refcount`` for persistent strings can be replaced with map ptr offset to cache class lookups by name
+
+map_ptr
+-------
+
+As mentioned in the introduction, SHM is shared for all child processes. Thus, no changes can be made to the data
+structures living in SHM unless 1. the changes are supposed to be reflected in all child processes and 2. the SHM
+segment has been locked before any changes are written to avoid data races.
+
+Sometimes the data structures contain a field that should be different per process or request. The ``map_ptr`` mechanism
+can be used to achieve this. In short, instead of storing a pointer to per-process data directly (as that would affect
+all processes but each having different addresses for this process specific data), a unique offset is assigned during
+compilation. During runtime, local memory is allocated to hold enough space for all the entries referenced in these
+offsets. Each offset then corresponds to an item in the local memory segment without needing to know the per-process
+location of the exact element.
+
+The map ptr can also store real pointers. This can be useful when the data structure doesn't live in SHM and thus
+doesn't need to store the value in a separate map. The fact that pointers are aligned by their size is used to
```

Comment:
I think this is actually wrong. We're not just storing pointers to pointers but also pointers to other data structures. I think `ZEND_MAP_PTR_DEF(bool, foo)` would fail as its alignment is 1 byte and it thus could be stored in 0xABC1 without packing. So we're relying on the fact that the structures we're pointing themselves to have an alignment of >1. Maybe `ZEND_MAP_PTR_DEF` could even assert that.

@nikic Can you confirm this?

Reply:
Generally yes, with the caveat that there are both type alignment and allocator alignment guarantees. E.g. all ZMM allocations are 4/8 byte aligned, and the system allocator has similar guarantees. You couldn't store a pointer into the middle of a string though, which might not be sufficiently aligned.